### PR TITLE
#233 fix SA when disconnecting network models

### LINKED ISSUE #233: Store all SVarCs and generators in inputs even if they will receive default model
The agreed validation scheme of contingencies in the security analysis simulation requires that all network elements that could be reference in a contingency are present in the DFL inputs. 

Currently, Static Var Compensators and generators that will not receive a specific dynamic model in the DYD are not copied in the DFL inputs data structures.

These elements should be stored in the DFL inputs, with a flag indicating that no specific model will be assigned (a default model will be assigned by Dynawo).

### PR DESCRIPTION
**Pull request by marcochiaramello**
Closes #233

### CODE REVIEW

File: sources/Algo/include/AlgorithmsResults.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ */
+class AlgorithmsResults {
+ public:
+  /**
+   * @brief Construct a new Algorithms Results object
+   *
+   */
+  AlgorithmsResults() : isAtLeastOneGeneratorRegulating(false) {}
+
+  bool isAtLeastOneGeneratorRegulating;  ///< boolean determining if at least a generator is regulating the voltage
```
Comment (cossartque) [inline sources/Algo/include/AlgorithmsResults.h:34]: at least one

File: sources/Algo/include/ContingencyValidationAlgorithm.h [hunk 2]
```diff
@@ -73,27 +87,54 @@ class ValidContingencies {
   ElementContingenciesMap elementContingencies_;        ///< For each element identifier, all the contingencies where it is referenced
   ValidatingContingenciesMap validatingContingencies_;  ///< All contingencies with valid elements found, indexed by contingencyId
   std::vector<ContingencyRef> validContingencies_;      ///< Only valid contingencies
+  ElementIds elementsNetworkType_;                      ///< set containing contingencies elements id using network model
```
Comment (cossartque) [inline sources/Algo/include/ContingencyValidationAlgorithm.h:90]: Set

File: sources/Algo/include/GeneratorDefinitionAlgorithm.h [hunk 3]
```diff
@@ -152,6 +164,14 @@ class GeneratorDefinitionAlgorithm {
    */
   bool IsOtherGeneratorConnectedBySwitches(const NodePtr& node) const;
 
+  /**
+   * @brief determine is targetP of generator is initially in diagram
```
Comment (cossartque) [inline sources/Algo/include/GeneratorDefinitionAlgorithm.h:168]: if targetP ?

File: sources/Context.h [hunk 4]
```diff
@@ -128,18 +131,18 @@ class Context {
   bool checkConnexity() const;
 
   /**
-   * @brief Register a callback to call at each node
+   * @brief Register a callback to call at each node of the main topological island
```
Comment (cossartque) [inline sources/Context.h:134]: main connected component ?

Comment (marcochiaramello) [reply sources/Context.h:134]: it is a synonym

File: sources/Algo/include/AlgorithmsResults.h (hunk 1 above)
Comment (marcochiaramello) [reply sources/Algo/include/AlgorithmsResults.h:34]: done, thx

**Review by cossartque — APPROVED** [review]

File: sources/Algo/src/GeneratorDefinitionAlgorithm.cpp [hunk 5]
```diff
@@ -37,38 +37,44 @@ GeneratorDefinitionAlgorithm::operator()(const NodePtr& node) {
...
-        } else {
-          model = useInfiniteReactivelimits_ ? GeneratorDefinition::ModelType::REMOTE_SIGNALN : GeneratorDefinition::ModelType::REMOTE_DIAGRAM_PQ_SIGNALN;
-        }
-        break;
-      case dfl::inputs::NetworkManager::NbOfRegulating::MULTIPLES:
+    GeneratorDefinition::ModelType model = GeneratorDefinition::ModelType::NETWORK;
+    if (isTargetPValid(generator) && generator.isVoltageRegulationOn) {
+      auto it = busMap_.find(generator.regulatedBusId);
+      assert(it != busMap_.end());
+      auto nbOfRegulatingGenerators = it->second;
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/GeneratorDefinitionAlgorithm.cpp:44]: I do agree with you on both comments, better to be explicit when possible, making code easier to be read

[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/GeneratorDefinitionAlgorithm.cpp]: done

File: sources/Algo/src/SVarCDefinitionAlgorithm.cpp [hunk 6]
```diff
@@ -18,20 +18,23 @@ namespace algo {
...
-        model = hasRemoteRegulation ? ModelType::SVARCPVPROPREMOTEMODEHANDLING : ModelType::SVARCPVPROPMODEHANDLING;
-      }
-    } else {
-      model = hasRemoteRegulation ? ModelType::SVARCPVREMOTE : ModelType::SVARCPV;
-      if (svarc.hasVoltagePerReactivePowerControl && !DYN::doubleIsZero(svarc.slope)) {
-        model = hasRemoteRegulation ? ModelType::SVARCPVPROPREMOTE : ModelType::SVARCPVPROP;
+    ModelType model = ModelType::NETWORK;
+    if (svarc.isRegulatingVoltage) {
+      model = ModelType::SVARCPV;
+      auto hasRemoteRegulation = (svarc.connectedBusId != svarc.regulatedBusId);
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/SVarCDefinitionAlgorithm.cpp]: done

**Review by menezesian — APPROVED** [review]

File: sources/Algo/src/ContingencyValidationAlgorithm.cpp [hunk 7]
```diff
@@ -17,43 +17,34 @@
 namespace dfl {
 namespace algo {
 void
-ContingencyValidationAlgorithm::operator()(const NodePtr& node) {
+ContingencyValidationAlgorithmOnNodes::operator()(const NodePtr& node, std::shared_ptr<AlgorithmsResults>&) {
   using Type = inputs::ContingencyElement::Type;
 
   for (const auto& line : node->lines) {
-    validContingencies_.markElementValid(line.lock()->id, Type::LINE);
+    validContingencies_.markElementValid(line.lock()->id, Type::LINE, false);
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/ContingencyValidationAlgorithm.cpp]: done

File: sources/Algo/src/ContingencyValidationAlgorithm.cpp [hunk 8]
```diff
@@ -81,6 +72,9 @@ ValidContingencies::markElementValid(const ElementId& elementId, inputs::Conting
       if (contingencyElementIt != contingency.elements.end() && inputs::ContingencyElement::isCompatible((*contingencyElementIt).type, elementType)) {
         // If type is compatible, add the element to the list of valid elements found for the contingency
         validatingContingencies_[contingency.id].insert(elementId);
+        if (isNetwork) {
+          elementsNetworkType_.insert(elementId);
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply sources/Algo/src/ContingencyValidationAlgorithm.cpp]: done

File: tests/algo/TestAlgo.cpp [hunk 9]
```diff
@@ -314,40 +314,44 @@ TEST(Generators, base) {
...
+  std::shared_ptr<dfl::algo::AlgorithmsResults> algoRes(new dfl::algo::AlgorithmsResults());
+  for (const auto& node : nodes) {
+    algo_infinite(node, algoRes);
+  }
 
-  std::for_each(nodes.begin(), nodes.end(), algo_infinite);
-
-  ASSERT_EQ(4, generators.size());
+  ASSERT_EQ(5, generators.size());
+  ASSERT_TRUE(algoRes->isAtLeastOneGeneratorRegulating);
```
[the remark this answers is not in the document]
Comment (marcochiaramello) [reply tests/algo/TestAlgo.cpp:354]: added a test

File: sources/Algo/include/ContingencyValidationAlgorithm.h [hunk 10]
```diff
@@ -113,6 +113,7 @@ class ContingencyValidationAlgorithmOnNodes {
 
  private:
   ValidContingencies& validContingencies_;  ///< the contingencies being validated by the algorithm
+  const bool notInNetwork_ = false;         ///< boolean determining the component is not modelled as network
```
Comment (rosiereflo) [inline sources/Algo/include/ContingencyValidationAlgorithm.h:116]: No need to put it as attribute, a local variable is fine

**Review by rosiereflo — CHANGES_REQUESTED** [review]

File: sources/Algo/include/ContingencyValidationAlgorithm.h (hunk 10 above)
Comment (marcochiaramello) [reply sources/Algo/include/ContingencyValidationAlgorithm.h:116]: ok!

**Review by rosiereflo — APPROVED** [review]

**Pull request merged**